src/shims/unix/fd.rs

```diff
@@ -394,7 +394,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         match result {
             Ok(read_bytes) => {
                 // If reading to `bytes` did not fail, we write those bytes to the buffer.
-                this.write_bytes_ptr(buf, bytes)?;
+                this.write_bytes_ptr(buf, bytes[..usize::try_from(read_bytes).unwrap()].to_vec())?;
```

Comment:
```suggestion
                // Crucially, if fewer than `bytes.len()` bytes were read, only write
                // that much into the output buffer!
                this.write_bytes_ptr(buf, bytes[..usize::try_from(read_bytes).unwrap()].to_vec())?;
```
Also, `to_vec` should not be needed. `.iter().copied()` should work.